libs/dal/src/shared/cache/cache.service.ts

```diff
@@ -0,0 +1,83 @@
+import Redis from 'ioredis';
+
+export interface ICacheService {
+  set(key: string, value: string, options?: CachingConfig);
+  get(key: string);
+  del(key: string);
+  delByPattern(pattern: string);
+}
+
+export type CachingConfig = {
+  ttl?: number;
+};
+
+export class CacheService implements ICacheService {
+  private readonly client: Redis;
+  private readonly DEFAULT_TTL = 60;
+  private readonly cacheTtl: number;
+
+  constructor(
+    private config: {
+      cachePort: string;
+      cacheHost: string;
+      cachePassword?: string;
+      cacheConnectTimeout?: string;
+      cacheKeepAlive?: string;
+      cacheFamily?: string;
+      cacheKeyPrefix?: string;
+      cacheTtl?: string;
+    }
+  ) {
+    this.client = new Redis(Number(this.config.cachePort), this.config.cacheHost, {
+      password: this.config.cachePassword,
+      connectTimeout: Number(this.config.cacheConnectTimeout) ?? 50000,
+      keepAlive: Number(this.config.cacheKeepAlive) ?? 30000,
+      family: Number(this.config.cacheFamily) ?? 4,
+      keyPrefix: this.config.cacheKeyPrefix ?? '',
+    });
+    this.cacheTtl = Number(this.config.cacheTtl) ?? this.DEFAULT_TTL;
+  }
+
+  public async set(key: string, value: string, options?: CachingConfig) {
+    this.client.set(key, value);
+    this.updateTtl(key, options);
+  }
+
+  public async keys(pattern?: string) {
+    const ALL_KEYS = '*';
+    const queryPattern = pattern ?? ALL_KEYS;
+
+    return this.client.keys(queryPattern);
+  }
+
+  public async get(key: string) {
+    return this.client.get(key);
+  }
+
+  public async del(key: string) {
+    return this.client.del([key]);
+  }
+
+  public async delByPattern(pattern: string) {
+    const client = this.client;
+    const stream = client.scanStream({
+      match: pattern,
+    });
+
+    stream.on('data', function (keys) {
+      if (keys.length) {
+        const pipeline = client.pipeline();
+        keys.forEach(function (key) {
+          pipeline.del(key);
+        });
+        pipeline.exec();
+      }
+    });
+  }
```

Comment:
this one doesn't need to be async, or if you want to resolve it after `pipeline.exec()` then you need to convert it to the promise and resolve it after exec, don't forget about the rejection then

Reply:
No in my opinion this one should not be awaited. because the exaction of the usecase is not dependent on its result. WDYT? @LetItRock